### PR TITLE
feat: add Kotlin android sample

### PR DESCRIPTION
**Pull request by qingzhuozhen**
### Summary

- add Kotlin android sample
- add some comments
- modify a bit for compatibility

### Checklist

* [x] Does your PR title have the correct [title format](https://github.com/amplitude/Amplitude-Kotlin/blob/main/CONTRIBUTING.md#pr-commit-title-conventions)?
* Does your PR have a breaking change?:  no

### CODE REVIEW

File: core/src/main/java/com/amplitude/core/Amplitude.kt [hunk 1]
```diff
@@ -68,104 +73,131 @@ open class Amplitude internal constructor(
...
      */
-    fun setGroup(groupType: String, groupName: String, options: EventOptions? = null) {
+    fun setGroup(groupType: String, groupName: String, options: EventOptions? = null): Amplitude {
         val identify = Identify().set(groupType, groupName)
         identify(identify, options)
+        return this
     }
 
     /**
      * ets the user's groups.
```
Comment (haoliu-amp) [inline core/src/main/java/com/amplitude/core/Amplitude.kt:164]: typo

File: core/src/main/java/com/amplitude/core/Amplitude.kt [hunk 2]
```diff
@@ -68,104 +73,131 @@ open class Amplitude internal constructor(
...
         event.groupProperties = identify.properties
         options ?. let {
             event.mergeEventOptions(it)
         }
         process(event)
+        return this
     }
 
     /**
      * Sets the user's group.
```
Comment (haoliu-amp) [inline core/src/main/java/com/amplitude/core/Amplitude.kt:155]: Set

File: core/src/main/java/com/amplitude/core/utilities/JSONUtil.kt [hunk 3]
```diff
@@ -157,50 +157,50 @@ internal fun JSONArray.toIntArray(): IntArray {
...
+    event.idfv = this.optionalString("idfv", null)
+    event.adid = this.optionalString("adid", null)
+    event.androidId = this.optionalString("android_id", null)
     event.eventId = if (this.has("event_id")) this.getInt("event_id") else null
     event.sessionId = this.getLong("session_id")
-    event.insertId = this.optString("insert_id", null)
+    event.insertId = this.optionalString("insert_id", null)
     event.library = if (this.has("library")) this.getString("library") else null
-    event.partnerId = this.optString("partner_id", null)
+    event.partnerId = this.optionalString("partner_id", null)
```
Comment (haoliu-amp) [inline core/src/main/java/com/amplitude/core/utilities/JSONUtil.kt:195]: do we need app set id too this time?

**Review by haoliu-amp — CHANGES_REQUESTED** [review]

File: core/src/main/java/com/amplitude/core/Amplitude.kt [hunk 4]
```diff
@@ -68,104 +73,131 @@ open class Amplitude internal constructor(
...
         callback ?. let {
             event.callback = it
         }
         process(event)
+        return this
     }
 
-    fun track(eventType: String, eventProperties: JSONObject? = null, options: EventOptions? = null) {
+    /**
+     * Log event with the specified event type, event properties, and optional event options
```
Comment (haoliu-amp) [inline core/src/main/java/com/amplitude/core/Amplitude.kt:92]: Log an event.

Q: Why not we fully leverage javadoc? Like using @param ?

File: core/src/main/java/com/amplitude/core/Amplitude.kt [hunk 5]
```diff
@@ -29,6 +29,11 @@ import kotlinx.coroutines.launch
 import org.json.JSONObject
 import java.util.concurrent.Executors
 
+/**
+ * <h1>Amplitude</h1>
+ * This is the SDK instance class that contains all of the SDK functionality.<br><br>
+ * Many of the SDK functions return the SDK instance back, allowing you to chain multiple method calls together
```
Comment (haoliu-amp) [inline core/src/main/java/com/amplitude/core/Amplitude.kt:35]: There're multiple places not having period for a sentence. Please add them.

Comment (haoliu-amp) [inline core/src/main/java/com/amplitude/core/Amplitude.kt:35]: Plural - chain multiple methods

File: core/src/main/java/com/amplitude/core/Amplitude.kt [hunk 6]
```diff
@@ -68,104 +73,131 @@ open class Amplitude internal constructor(
...
     @Deprecated("Please use 'track' instead.", ReplaceWith("track"))
-    fun logEvent(event: BaseEvent) {
-        track(event)
+    fun logEvent(event: BaseEvent): Amplitude {
+        return track(event)
     }
 
-    fun track(event: BaseEvent, callback: EventCallBack? = null) {
+    /**
+     * Track an event
```
Comment (haoliu-amp) [inline core/src/main/java/com/amplitude/core/Amplitude.kt:81]: period

File: core/src/main/java/com/amplitude/core/Amplitude.kt [hunk 7]
```diff
@@ -68,104 +73,131 @@ open class Amplitude internal constructor(
...
         event.eventProperties = eventProperties
         options ?. let {
             event.mergeEventOptions(it)
         }
         process(event)
+        return this
     }
 
     /**
      * Identify. Use this to send an Identify object containing user property operations to Amplitude server.
```
Comment (haoliu-amp) [inline core/src/main/java/com/amplitude/core/Amplitude.kt:106]: Should we explain what exactly this does? 
Like - 
You can modify user properties by calling this API.

**Review by haoliu-amp — CHANGES_REQUESTED** [review]

File: core/src/main/java/com/amplitude/core/Amplitude.kt (hunk 4 above)
Comment (qingzhuozhen) [reply core/src/main/java/com/amplitude/core/Amplitude.kt:92]: We can add param if needed, the only reason for not doing that earlier is https://kotlinlang.org/docs/coding-conventions.html#documentation-comments 

I can add these to account for java style, kdoc still supports them

File: core/src/main/java/com/amplitude/core/utilities/JSONUtil.kt (hunk 3 above)
Comment (qingzhuozhen) [reply core/src/main/java/com/amplitude/core/utilities/JSONUtil.kt:195]: Good catch, missed that when just looking at the http api docs.. We should add that into the api doc

File: core/src/main/java/com/amplitude/core/Amplitude.kt (hunk 7 above)
Comment (qingzhuozhen) [reply core/src/main/java/com/amplitude/core/Amplitude.kt]: Add that in the comments

File: samples/kotlin-android-app/src/main/java/com/amplitude/android/sample/MainApplication.kt [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+            Configuration(
+                apiKey = AMPLITUDE_API_KEY,
+                context = applicationContext
+            )
+        )
+
+        // add sample plugin
+        amplitude.add(object : Plugin {
+            override val type: Plugin.Type = Plugin.Type.Enrichment
+            override lateinit var amplitude: com.amplitude.core.Amplitude
```
Comment (justin-fiedler) [inline samples/kotlin-android-app/src/main/java/com/amplitude/android/sample/MainApplication.kt:30]: Is the `lateinit amplitude` override necessary?

File: samples/kotlin-android-app/src/main/java/com/amplitude/android/sample/MainActivity.kt [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+            .set("custom-properties", "sample")
+        val options = EventOptions()
+        options.plan = Plan(branch = "test")
+        amplitude.identify(identify, options)
+
+        // set groups fro this user
+        amplitude.setGroup("test-group-type", "android-kotlin-sample")
+        amplitude.setGroup("orgId", "15")
+        amplitude.setGroup("sport", arrayOf("tennis", "soccer")) // list values
+
```
Comment (justin-fiedler) [inline samples/kotlin-android-app/src/main/java/com/amplitude/android/sample/MainActivity.kt:30]: groupIdentify?

**Review by justin-fiedler — CHANGES_REQUESTED** [review]

File: samples/kotlin-android-app/src/main/java/com/amplitude/android/sample/MainActivity.kt (hunk 9 above)
Comment (qingzhuozhen) [reply samples/kotlin-android-app/src/main/java/com/amplitude/android/sample/MainActivity.kt:30]: Good catch! Let me add example for groupIdentify

File: samples/kotlin-android-app/src/main/java/com/amplitude/android/sample/MainApplication.kt (hunk 8 above)
Comment (qingzhuozhen) [reply samples/kotlin-android-app/src/main/java/com/amplitude/android/sample/MainApplication.kt:30]: That is still needed, there is a `var amplitude` inside Plugin interface, and the amplitude instance is not filled until the setup call.

File: core/src/main/java/com/amplitude/core/Amplitude.kt [hunk 10]
```diff
@@ -68,104 +73,174 @@ open class Amplitude internal constructor(
...
         this.idContainer.identityManager.editIdentity().setUserId(userId).commit()
+        return this
     }
 
-    fun setDeviceId(deviceId: String) {
+    /**
+     * Sets a custom device id. <b>Note: only do this if you know what you are doing!</b>
+     *
+     * @param deviceId custom device id
+     * @return the Amplitude
```
Comment (haoliu-amp) [inline core/src/main/java/com/amplitude/core/Amplitude.kt:147]: the Amplitude instance?
Same for others

Comment (qingzhuozhen) [reply core/src/main/java/com/amplitude/core/Amplitude.kt]: Sounds good, already updated

**Review by justin-fiedler — APPROVED** [review]

**Pull request merged**